arangod/Pregel/Algos/AIR/Greenspun/Interpreter.cpp

```diff
@@ -0,0 +1,810 @@
+        });
+      }
+
+      if (auto res = frame.setVariable(nameSlice.copyString(), builder.slice()); res.fail()) {
+        return res;
+      }
+    } else {
+      return EvalError("expected pair at position " + std::to_string(iter.index()) +
+                       " at list of bindings, found: " + pair.toJson());
+    }
+  }
+
+  StackFrameGuard<true> guard(ctx, std::move(frame));
+
+  // Now do a seq evaluation of the remaining parameter
+  return SpecialSeq(ctx, paramIterator, result).mapError([](EvalError& err) {
+    err.wrapMessage("in evaluation of let-statement");
+  });
+}
+
+EvalResult Evaluate(Machine& ctx, ArrayIterator paramIterator, VPackBuilder& result);
+
+EvalResult SpecialQuasiQuoteInternal(Machine& ctx, ArrayIterator other, VPackBuilder& result) {
+  if (other.valid()) {
+    Slice first = *other;
+    if (first.isString() && first.isEqualString("unquote")) {
+      ++other;
+      if (!other.valid() || !other.isLast()) {
+        return EvalError("expected one parameter for unquote");
+      }
+      return Evaluate(ctx, *other, result);
+    } else if (first.isString() && first.isEqualString("unquote-splice")) {
+      ++other;
+      if (!other.valid() || !other.isLast()) {
+        return EvalError("expected one parameter for unquote");
+      }
+      VPackBuilder tempResult;
+      if (auto res = Evaluate(ctx, *other, tempResult); res.fail()) {
+        return res;
+      }
+      auto tempSlice = tempResult.slice();
+      if (tempSlice.isArray()) {
+        result.add(VPackArrayIterator(tempSlice));
+      } else {
+        result.add(tempSlice);
+      }
+      return {};
+    }
+  }
+
+  {
+    VPackArrayBuilder ab(&result);
+
+    for (; other.valid(); ++other) {
+      auto&& part = *other;
+      if (part.isArray()) {
+        if (auto res = SpecialQuasiQuoteInternal(ctx, VPackArrayIterator(part), result);
+            res.fail()) {
+          return res;
+        }
+      } else {
+        result.add(part);
+      }
+    }
+  }
+
+  return {};
+}
+
+EvalResult SpecialQuasiQuote(Machine& ctx, ArrayIterator paramIterator, VPackBuilder& result) {
+  if (!paramIterator.valid()) {
+    return EvalError("quasi-quote expects one parameter");
+  }
+
+  auto value = *paramIterator++;
+  if (paramIterator.valid()) {
+    return EvalError("Excess elements in quasi-quote call");
+  }
+  if (value.isArray()) {
+    return SpecialQuasiQuoteInternal(ctx, ArrayIterator(value), result);
+  }
+
+  result.add(value);
+  return {};
+}
+
+EvalResult SpecialStr(Machine& ctx, ArrayIterator paramIterator, VPackBuilder& result) {
+  std::stringstream ss;
+
+  for (; paramIterator.valid(); ++paramIterator) {
+    if ((*paramIterator).isString()) {
+      ss << (*paramIterator).copyString();
+    } else {
+      return EvalError(std::string{"`str` expecting string, not "} +
+                       (*paramIterator).typeName());
+    }
+  }
+
+  result.add(VPackValue(ss.str()));
+  return {};
+}
```

Comment:
`str` is a special form. It does not evaluate its parameters. So `["str", ["my-func"]]` will fail with `expected string` and `my-func` is never called.

Reply:
And why was it added: because markus wanted to replace the current semantic of `"xyz"` being the string value `xyz` with resolving this to the variable with name `xyz`. To then write a string you would have to use `["str", "my string"]`.